### ISSUE TITLE
[대회 중간 점검] [Day 5] 병현이의 인사이트 회고 및 중간 점검

### ISSUE THREAD

**Issue by Bkankim** [AIBootcamp14/computervisioncompetition-cv-1#26]
# CV 문서분류 대회 핵심 인사이트 정리

## 🎯 성과 요약
* **최종 성능**: F1 0.934 (베이스라인 0.1659 대비 5.6배 향상)
* **로컬 CV**: 0.9475 ± 0.0076 (5-fold)
* **CV-제출 갭**: 1.3%p (초기 19%p에서 대폭 축소)

### 1. 모델 선택의 중요성
* **ResNet34 → Swin-B**: 단순 모델 교체만으로도 극적 성능 향상
* **해상도 확대**: 32px → 384px가 문서 인식에 결정적

### 2. 증강 전략의 진화
* **문서 특화 증강**: 90도 회전이 랜덤 회전보다 효과적
* **적응형 Hard Aug**: 에포크별 20%→50% 점진 증가가 과적합 방지
* **테스트 특화**: 블러, 노이즈, JPEG 압축으로 실제 테스트 환경 모사

### 3. CV-제출 갭 해결 과정

```
초기: CV 0.93 vs 제출 0.74 (19%p 갭)
↓
적응형 증강 적용
↓
최종: CV 0.9475 vs 제출 0.911 (3.6%p 갭)
```

### 4. 앙상블의 파워
* **5-Fold + 5-TTA**: 25개 예측 조합으로 안정성 확보
* **소프트맥스 앙상블**: 하드 voting보다 확률 평균이 우수

### 5. 효율성 최적화
* **TTA 선별**: 15개→5개 변형으로 3배 속도 향상
* **Mixed Precision**: 메모리 효율성과 속도 개선
* **추론 시간**: 4시간→13분으로 단축

### 성능 진화 단계별 기여도
1. **백본 + 하이퍼파라미터**: F1 +0.25 (0.17→0.42)
2. **TTA + 정규화**: F1 +0.21 (0.42→0.63)
3. **5-fold + 스케줄러**: F1 +0.11 (0.63→0.74)
4. **앙상블 + 패딩 + 캘리브레이션**: F1 +0.13 (0.74→0.87)
5. **적응형 증강 + Swin-B**: F1 +0.041 (0.87→0.911)

### 검증된 핵심 기법 Top 5
1. **모델 업그레이드** (ResNet34→Swin-B): 기본 성능의 기반
2. **해상도 전략** (32→384px + 비율보존): 문서 인식 정확도 향상
3. **5-Fold 앙상블**: 과적합 방지와 안정성 확보
4. **도메인 적응 증강**: CV-제출 갭 해소의 핵심
5. **TTA 최적화**: 성능과 효율성의 균형점

### 실패한 시도들
* **Pseudo Labeling**: 노이즈 데이터로 인한 성능 저하 + 대회 룰에 위배
* **과도한 Label Smoothing** (0.2): 언더피팅 유발
* **Hard Validation 분리**: 학습 데이터 부족으로 역효과 (엄청난 과적합)

## 📊 클래스별 분석 결과
* **최고 성능**: 클래스 1, 8, 16 (95%+ 정확도)
* **취약 클래스**: 클래스 7(72%), 3(74%), 14(76%)
* **개선 방향**: 취약 클래스 대상 targeted augmentation 필요 (단순 서류 형식이므로, 텍스트 임베딩도 효과적일 것이라 생각)

## 🎲 재현성 보장 요소
* **시드 고정**: 42로 통일
* **결정론적 설정**: `torch.backends.cudnn.benchmark = True`
* **Stratified K-Fold**: 클래스 비율 유지
* **모든 실험 추적**: 설정-성능 매핑 완료

## 🚀 최종 성공 공식

```
Swin-B 384px + 5-Fold CV + 적응형 Hard Aug + 5-TTA = F1 0.911
```

**핵심 하이퍼파라미터:**
- Hard Aug 확률: 20%→50% (에포크별 점진 증가)  
- Label Smoothing: 0.1 (과정규화 완화)
- 추론 시간: 13분 15초 (효율성 확보)

### 로컬-제출 갭 원인 진단
**초기 상황**: 로컬 F1 0.93 vs 제출 F1 0.87 (6%p 갭)
* **과적합 (Overfitting)**: 한정된 데이터 암기로 편향된 학습
* **과신 (Overconfidence)**: 예측 확률이 정답률보다 높음, 확률 보정 어려움

### 검증된 갭 해소 전략
1. **혼합 증강 + 확률 스위치**: 에포크별 Hard Aug 20%→50% 스케줄링
2. **도메인 시프트 대응**: Train-Test 품질 괴리가 주요 원인으로 파악
3. **점증적 오염 학습**: 오염 데이터 예측 강화를 위한 확률적 증강

### 실험 결과 요약
* **Label Smoothing 0.2→0.1**: 로컬 성능 유지, 갭 축소 제한적
* **EfficientNetV2-B3 업그레이드**: 속도 향상 + 0.6% 성능 개선
* **적응형 증강 시스템**: **핵심 성공 요인** - 제출 점수 4% 상승, 갭 절반 축소

### 오답률 높은 클래스 (Target Classes)
1. **입퇴원확인서** (클래스 7): 72% 정확도
2. **외래진료확인서** (클래스 3): 74% 정확도  
3. **의견서** (클래스 14): 76% 정확도

**공통 패턴**: 글자 위주 서류에서 취약함 → OCR/텍스트 파싱 보강 고려

### 향후 개선 계획
* **SimCLR ResNet**: fine tuning 및 백본 앙상블 검토
* **메모리 최적화**: 배치 크기 증대 → 거대 모델 활용 가능
* **서버 맞춤 최적화**: 하드웨어 스펙 기반 백본-하이퍼파라미터 조합 리서치
* **Targeted Augmentation**: 취약 클래스 전용 증강 전략

## 💭 교훈
1. **기본기가 80%**: 좋은 모델 + 적절한 해상도
2. **도메인 지식 활용**: 문서 특성 반영한 증강 전략
3. **체계적 접근**: 가설→실험→검증→개선 사이클
4. **갭 해소 우선**: 로컬-제출 일치도가 실험 자유도 결정
5. **효율성 고려**: 성능과 속도의 균형점 찾기
6. **도메인 시프트 인식**: Train-Test 품질 차이가 핵심 문제
7. **점진적 강화 학습**: 적응형 증강이 과적합 방지의 열쇠

**Comment by Bkankim** [comment]
최종 최고 성능을 보인 Swin-B 백본 모델은 Transformer 기반 모델이기 때문에
Transformer 기반 모델 Task를 진행중이신
@Seowoo-C  @joy007fun  (정서우, 임예슬) 님들에게 금일 오전 중으로 인사이트 공유드릴 예정입니다!

**Comment by nimnusmik** [comment]
정리 너무 잘하셨네요........ 최고입니다

**Comment by nimnusmik** [comment]
그런데 기준님은 왜 태그하셨는지........?!?!?!?

**Comment by Bkankim** [comment]
어이쿠, 실수..

**Comment by Seowoo-C** [comment]
정말 많은 인사이트 얻어갑니다. 공유 감사합니다!! 
확실히 가장 성능이 좋았으니까 swin-b로도 한번 도전해봐야겠어요!!
